File: src/main/java/org/blacksun/graph/node/NamedGraphNode.java
package org.blacksun.graph.node;

import org.blacksun.graph.channel.Channel;
import org.blacksun.graph.channel.ChannelFactory;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

public class NamedGraphNode implements GraphNode {
    private String name;
    private final ChannelFactory factory;
    private List<Channel> connections;
    private boolean selected;

    NamedGraphNode(@NotNull String name, @NotNull ChannelFactory factory) {
        this.name = name;
        this.factory = factory;
        connections = new ArrayList<>();
    }

    public String getName() {
        return name;
    }

    public void setName(@NotNull String name) {
        this.name = name;
    }

    @Override
    public boolean isSelected() {
        return selected;
    }

    @Override
    public void setSelected(boolean selected) {
        this.selected = selected;
    }

    @Override
    public List<GraphNode> getConnectedNodes() {
        return connections.stream()
                .map(Channel::getToNode)
                .collect(Collectors.toList());
    }

    @Override
    public List<Channel> getConnections() {
        return connections;
    }

    @Override
    public boolean isConnected(@NotNull GraphNode node) {
        return connections.stream().anyMatch(ch -> ch.getToNode().equals(node));
    }

    @Override
    public Channel getConnection(@NotNull GraphNode node) {
        return connections.stream()
                .filter(ch -> ch.getToNode().equals(node))
                .findAny()
                .orElseThrow(() -> new RuntimeException(node + " is not connected"));
    }

    @Override
    public Channel addConnection(@NotNull Channel channel) {
        connections.add(channel);
        return channel;
    }

    @Override
    public Channel addConnectedNode(@NotNull GraphNode node, int weight) {
        Channel channel = factory.createChannel(this, node, weight);
        channel.connect();
        return channel;
    }

    @Override
    public void removeConnection(@NotNull Channel channel) {
        connections.remove(channel);
    }

    @Override
    public void removeConnectedNode(@NotNull GraphNode node) {
        connections.stream()
                .filter(ch -> ch.getToNode().equals(node))
                .findAny()
                .ifPresent(Channel::remove);
    }

    @Override
    public int getOrder() {
        return connections.size();
    }

    @Override
    public String stringRepresentation() {
        StringBuilder sb = new StringBuilder();
        sb.append(name).append(" [\n");
        connections.forEach(ch -> sb.append("\t")
                .append(ch.stringRepresentation())
                .append("\n"));
        sb.append("]");
        return sb.toString();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o)
            return true;
        if (o == null || getClass() != o.getClass())
            return false;
        NamedGraphNode that = (NamedGraphNode) o;
        return Objects.equals(name, that.name) &&
                toString().equals(that.toString()); // && Objects.equals(connections, that.connections);
    }

    @Override
    public int hashCode() {
        return Objects.hash(name, toString()); //, connections);
    }

    @Override
    public String toString() {
        return "Node[name=" + name + ", order=" + getOrder() + "]";
    }
}

File: src/main/java/org/blacksun/graph/node/NamedGraphNode.kt
package org.blacksun.graph.node

import org.blacksun.graph.channel.Channel
import org.blacksun.graph.channel.ChannelFactory

import java.util.ArrayList
import java.util.Objects

open class NamedGraphNode internal constructor(var name: String, private val factory: ChannelFactory) : GraphNode {
    override val connections = ArrayList<Channel>()
    override var isSelected = false
    override var isTerminal = false

    override val connectedNodes: List<GraphNode>
        get() = connections.map { it.toNode }

    override val order: Int
        get() = connections.size

    override fun isConnected(node: GraphNode): Boolean {
        return connections.stream().anyMatch { ch -> ch.toNode == node }
    }

    override fun getConnection(node: GraphNode): Channel {
        return connections.stream()
                .filter { ch -> ch.toNode == node }
                .findAny()
                .orElseThrow { RuntimeException(node.toString() + " is not connected") }
    }

    override fun addConnection(channel: Channel): Channel {
        connections.add(channel)
        return channel
    }

    override fun addConnectedNode(node: GraphNode, weight: Int): Channel {
        return factory.createChannel(this, node, weight)
    }

    override fun removeConnection(channel: Channel) {
        connections.remove(channel)
    }

    override fun removeConnectedNode(node: GraphNode) {
        connections.stream()
                .filter { ch -> ch.toNode == node }
                .findAny()
                .ifPresent { it.remove() }
    }

    override fun stringRepresentation(): String {
        val sb = StringBuilder()
        sb.append(name).append(" [\n")
        connections.forEach { ch ->
            sb.append("\t")
                    .append(ch.stringRepresentation())
                    .append("\n")
        }
        sb.append("]")
        return sb.toString()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other)
            return true
        if (other == null || javaClass != other.javaClass)
            return false
        val that = other as NamedGraphNode?
        return name == that!!.name && toString() == that.toString() // && Objects.equals(connections, that.connections);
    }

    override fun hashCode(): Int {
        return Objects.hash(name, toString()) //, connections);
    }

    override fun toString(): String {
        return "Node[name=$name, order=$order]"
    }
}
